src/mlte/store/underlying/http.py

```diff
@@ -0,0 +1,281 @@
+
+        return Namespace(**res.json())
+
+    def create_model(self, namespace_id: str, model: ModelCreate) -> Model:
+        url = f"{self.url}/api/namespace/{namespace_id}/model"
+        res = self.client.post(url, json=model.dict())
+        raise_for_response(res)
+
+        return Model(**res.json())
+
+    def read_model(self, namespace_id: str, model_id: str) -> Model:
+        url = f"{self.url}/api/namespace/{namespace_id}/model/{model_id}"
+        res = self.client.get(url)
+        raise_for_response(res)
+
+        return Model(**res.json())
+
+    def list_models(self, namespace_id: str) -> list[str]:
+        url = f"{self.url}/api/namespace/{namespace_id}/model"
+        res = self.client.get(url)
+        raise_for_response(res)
+
+        return typing.cast(list[str], res.json())
+
+    def delete_model(self, namespace_id: str, model_id: str) -> Model:
+        url = f"{self.url}/api/namespace/{namespace_id}/model/{model_id}"
+        res = self.client.delete(url)
+        raise_for_response(res)
+
+        return Model(**res.json())
+
+    def create_version(
+        self, namespace_id: str, model_id: str, version: VersionCreate
+    ) -> Version:
+        url = (
+            f"{self.url}/api/namespace/{namespace_id}/model/{model_id}/version"
+        )
+        res = self.client.post(url, json=version.dict())
+        raise_for_response(res)
+
+        return Version(**res.json())
+
+    def read_version(
+        self, namespace_id: str, model_id: str, version_id: str
+    ) -> Version:
+        url = f"{self.url}/api/namespace/{namespace_id}/model/{model_id}/version/{version_id}"
+        res = self.client.get(url)
+        raise_for_response(res)
+
+        return Version(**res.json())
+
+    def list_versions(self, namespace_id: str, model_id: str) -> list[str]:
+        url = (
+            f"{self.url}/api/namespace/{namespace_id}/model/{model_id}/version"
+        )
+        res = self.client.get(url)
+        raise_for_response(res)
+
+        return typing.cast(list[str], res.json())
+
+    def delete_version(
+        self, namespace_id: str, model_id: str, version_id: str
+    ) -> Version:
+        url = f"{self.url}/api/namespace/{namespace_id}/model/{model_id}/version/{version_id}"
+        res = self.client.delete(url)
+        raise_for_response(res)
+
+        return Version(**res.json())
+
+    # -------------------------------------------------------------------------
+    # Negotiation Card
+    # -------------------------------------------------------------------------
+
+    def write_negotiation_card(
+        self,
+        namespace_id: str,
+        model_id: str,
+        version_id: str,
+        artifact: NegotiationCardModel,
+    ) -> NegotiationCardModel:
+        url = f"{_url(self.url, namespace_id, model_id, version_id)}/negotiation-card"
+        res = self.client.post(url, json=artifact.dict())
+        raise_for_response(res)
+
+        return NegotiationCardModel(**res.json())
```

Comment:
Related to my previous comment, do we really need write/read/delete methods for each artifact separately? Can't we have a generic write/read/delete artifact method, that maybe. receives the model type and gets additional info, like URL, as needed? This looks like it will require a lot of almost-the-same code for each artifact, with only very minor difference (basically the model type and the actual URL. I like trying to avoid duplicating samey code if possible to make maintenance easier.

Reply:
See refactor.